src/main/java/com/google/devtools/build/lib/actions/FileArtifactValue.java

```diff
@@ -533,12 +555,100 @@ protected boolean couldBeModifiedByMetadata(FileArtifactValue o) {
     }
   }
 
+  private static final class TreeArtifactProxyFileArtifactValue extends FileArtifactValue {
```

Comment:
As discussed offline: I don't fully understand the reasons *why* this exists, so I don't think I can do better than the current javadoc.

Regarding the class name: it doesn't make sense to include `Symlink` in the name, since every tree artifact has one of these values associated with it, regardless of whether the `symlinkTargetExecPath` field is set.

(Maybe you meant separate classes for when the symlink is or isn't present? I'm not convinced that pulls its weight; unlike the `RemoteFileArtifactValue` case, there's no need to optimize for memory footprint here.)